File: api/src/main/java/net/asch/bulkit/api/capability/IDriveNetworkLink.java
package net.asch.bulkit.api.capability;

import net.minecraft.core.BlockPos;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;

import javax.annotation.Nullable;

public interface IDriveNetworkLink {
    int UNMAPPED_SLOT = -1;

    default void unmap(int slot) {
        map(slot, UNMAPPED_SLOT);
    }

    void map(int slot, int rootSlot);

    default void unlink(Player player) {
        linkTo(player, null);
    }

    void linkTo(Player player, @Nullable BlockPos blockPos);

    ItemStack disk(int slot);
}

File: api/src/main/kotlin/net/asch/bulkit/api/capability/IDriveNetworkLink.kt
package net.asch.bulkit.api.capability

import net.minecraft.core.BlockPos
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack

interface IDriveNetworkLink {
    fun unmap(slot: Int) = map(slot, UNMAPPED_SLOT)
    fun map(slot: Int, rootSlot: Int)

    fun unlink(player: Player) = linkTo(player, null)
    fun linkTo(player: Player, blockPos: BlockPos?)

    fun disk(slot: Int): ItemStack

    companion object {
        const val UNMAPPED_SLOT: Int = -1
    }
}